src/orders/components/OrderFulfillmentMetadataDialog/OrderFulfillmentMetadataDialog.tsx

```diff
@@ -0,0 +1,191 @@
+import { ChangeEvent } from "@dashboard/hooks/useForm";
+import { flattenErrors } from "@dashboard/utils/hook-form/errors";
+import { mapMetadataItemToInput } from "@dashboard/utils/maps";
+import { useEffect, useMemo } from "react";
+import { FieldArrayPath, FieldError, useFieldArray, useForm } from "react-hook-form";
+import { useIntl } from "react-intl";
+
+import { useHandleFulfillmentMetadataSubmit } from "./useHandleSubmit";
+
+export type FulfillmentMetadataDialogData = OrderDetailsFragment["fulfillments"][0];
+
+interface OrderFulfillmentMetadataDialogProps {
+  open: boolean;
+  onClose: () => void;
+  fulfillment: FulfillmentMetadataDialogData | undefined;
+}
+
+const getValidateMetadata =
+  (intl: ReturnType<typeof useIntl>) =>
+  (metadata: MetadataInput[]): true | string => {
+    const keys = metadata.map(entry => entry.key);
+    const uniqueKeys = new Set(keys);
+
+    if (uniqueKeys.size !== keys.length) {
+      return intl.formatMessage({
+        defaultMessage: "Metadata keys must be unique, remove duplicate key",
+        description: "metadata edit form, error message",
+        id: "MfWHGz",
+      });
+    }
+
+    if (keys.some(key => key === "")) {
+      return intl.formatMessage({
+        defaultMessage: "Metadata key cannot be empty",
+        description: "metadata edit form, error message",
+        id: "lb5uDM",
+      });
+    }
+
+    return true;
+  };
+
+const mapFieldArrayToMetadataInput = (
+  fields: Array<Record<"id", string> & MetadataInput>,
+): MetadataInput[] => {
+  return fields.map(field => ({
+    key: field.key,
+    value: field.value,
+  }));
+};
+
+export const OrderFulfillmentMetadataDialog = ({
+  onClose,
+  open,
+  fulfillment,
+}: OrderFulfillmentMetadataDialogProps) => {
+  const intl = useIntl();
+  const { onSubmit, lastSubmittedData, submitInProgress } = useHandleFulfillmentMetadataSubmit({
+    initialData: fulfillment,
+    onClose,
+  });
+
+  const formMethods = useForm<MetadataFormData>({
+    // Display last submitted data while re-fetching to avoid flicker on UI
+    values: submitInProgress
+      ? lastSubmittedData
+      : {
+          // Removes __typename from metadata item object
+          metadata: (fulfillment?.metadata ?? []).map(mapMetadataItemToInput),
+          privateMetadata: (fulfillment?.privateMetadata ?? []).map(mapMetadataItemToInput),
+        },
+  });
+
+  const { control, getValues, formState, trigger, reset } = formMethods;
+
+  // Metadata field arrays
+  const metadataControls = useFieldArray<MetadataFormData>({
+    control,
+    name: "metadata" as FieldArrayPath<MetadataFormData>,
+    rules: {
+      validate: getValidateMetadata(intl),
+    },
+  });
+
+  const privateMetadataControls = useFieldArray<MetadataFormData>({
+    control,
+    name: "privateMetadata" as FieldArrayPath<MetadataFormData>,
+    rules: {
+      validate: getValidateMetadata(intl),
+    },
+  });
```

Comment:
since this is all about metadata you may consider introducing abstraction on top, so we can eventually have:
1. MetadataDialog → ui
2. useMetadataDialogThings() → Hook with common stuff, validation etc (and tests)
3. OrderFulfillmentMetadataDialog → put things together, add custom stuff